federatedscope/core/fed_runner.py

```diff
@@ -427,3 +432,62 @@ def _handle_msg(self, msg, rcv=-1):
                 self.client[each_receiver].msg_handlers[msg.msg_type](msg)
                 self.client[each_receiver]._monitor.track_download_bytes(
                     download_bytes)
+
+    def check(self):
+        """
+        Check the completeness of Server and Client.
+
+        Returns:
+
+        """
+        if self.cfg.check_completeness:
+            try:
+                import os
+                import networkx as nx
+                import matplotlib.pyplot as plt
+                # Build check graph
+                G = nx.DiGraph()
+                flags = {0: 'Client', 1: 'Server'}
+                msg_handler_dicts = [
+                    self.client_class.get_msg_handler_dict(),
+                    self.server_class.get_msg_handler_dict()
+                ]
+                for flag, msg_handler_dict in zip(flags.keys(),
+                                                  msg_handler_dicts):
+                    role, oppo = flags[flag], flags[(flag + 1) % 2]
+                    for msg_in, (handler, msgs_out) in \
+                            msg_handler_dict.items():
+                        for msg_out in msgs_out:
+                            msg_in_key = f'{oppo}_{msg_in}'
+                            handler_key = f'{role}_{handler}'
+                            msg_out_key = f'{role}_{msg_out}'
+                            G.add_node(msg_in_key, subset=1)
+                            G.add_node(handler_key, subset=0 if flag else 2)
+                            G.add_node(msg_out_key, subset=1)
+                            G.add_edge(msg_in_key, handler_key)
+                            G.add_edge(handler_key, msg_out_key)
+                pos = nx.multipartite_layout(G)
+                plt.figure(figsize=(20, 15))
+                nx.draw(G,
+                        pos,
+                        with_labels=True,
+                        node_color='white',
+                        node_size=800)
+                fig_path = os.path.join(self.cfg.outdir, 'msg_handler.png')
+                plt.savefig(fig_path)
+                if nx.has_path(G, 'Client_join_in', 'Server_finish'):
+                    if nx.is_weakly_connected(G):
+                        logger.info(f'Completeness check passes! Save check '
+                                    f'results in {fig_path}.')
+                    else:
+                        logger.warning(
+                            f'Completeness check raises warning for '
+                            f'some handlers not in FL process! Save '
+                            f'check results in {fig_path}.')
+                else:
+                    logger.error(f'Completeness check fails for there is no'
+                                 f'path from `join_in` to `finish`! Save '
+                                 f'check results in {fig_path}.')
+            except Exception as error:
+                logger.warning(f'Completeness check failed for {error}!')
+        return
```

Comment:
so the reason is that we cannot say yes or not about the correctness/completeness, right?

Reply:
As the document shows, the correctness/completeness checks only check the message-handler pairs and raise three types of logs (info, warning, and error). If something goes wrong with Python code, we'd better keep the exception stack as it is. So the return value is meaningless.